File: JavaScript/Math-Array-String/66.js
// 3090. Maximum Length Substring With Two Occurrences


/**
 * @param {string} s
 * @return {number}
 */
var maximumLengthSubstring = function(s) {
    let charCount = {};
    let left = 0;
    let maxLength = 0;

    for (let right = 0; right < s.length; right++) {
        charCount[s[right]] = (charCount[s[right]] || 0) + 1;

        // Shrink the window if any character appears more than twice
        while (Object.values(charCount).some(count => count > 2)) {
            charCount[s[left]] -= 1;
            if (charCount[s[left]] === 0) {
                delete charCount[s[left]];
            }
            left++;
        }

        // Update the maximum length
        maxLength = Math.max(maxLength, right - left + 1);
    }

    return maxLength;
};

// Test cases
console.log(maximumLengthSubstring("bcbbbcba")); // Output: 4
console.log(maximumLengthSubstring("aaaa"));    // Output: 2

File: Python/Math-Array-String/66.py
# 3090. Maximum Length Substring With Two Occurrences


class Solution:
    def maximumLengthSubstring(self, s: str) -> int:
        from collections import defaultdict

        char_count = defaultdict(int)
        left = 0
        max_length = 0

        for right in range(len(s)):
            char_count[s[right]] += 1

            # Check if any character occurs more than twice
            while any(count > 2 for count in char_count.values()):
                char_count[s[left]] -= 1
                if char_count[s[left]] == 0:
                    del char_count[s[left]]
                left += 1

            max_length = max(max_length, right - left + 1)

        return max_length


sol = Solution()
output = sol.maximumLengthSubstring("bcbbbcba")
print(output) # Output: 4